In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
visa_df=pd.read_csv('Visadataset.csv')
visa_df

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

      case_status  
0          Denied  
1       Certified  
2          Denied  
3          Denied  
4       Certified  
...           ...  
25475   Certified  
25476   Certified  
25477   Certified  
25478   Certified  
25479   Certified  

[25480 rows x 12 columns]

In [2]:
cat_col=[key for key,value in visa_df.dtypes.items() if value=='object']
num_col=[key for key,value in visa_df.dtypes.items() if value!='object']
cat_col

['case_id',
 'continent',
 'education_of_employee',
 'has_job_experience',
 'requires_job_training',
 'region_of_employment',
 'unit_of_wage',
 'full_time_position',
 'case_status']

In [3]:
num_col

['no_of_employees', 'yr_of_estab', 'prevailing_wage']

**Encoding**

- encoding means converting categorical column data to numerical column data

- because ML models except the data should be in numerical format

- ML models developed by math algorithms

- so it is very important to apply encoding methods

- the following methods are

    - map
    
    - no.where
    
    - one hot encoding
    
    - label encodeer

**Map**

- mapping is one method to convert categorical values to numerical

- take one categorical column

- get the unique lables first

- make a dictonary with assign a number to each label

- for example case_status has 2 label

    - certified
    
    - denied

- assign 0 to certified and 1 to denied

- create a dictonary labels as keys and numbers as values only

- {'certified':0,'denied':1} 

In [4]:
visa_df=pd.read_csv('Visadataset.csv')
for j in cat_col:
    dict1={}
    for k,i in enumerate(visa_df[j].unique(),0):
        dict1[i]=k
    dict1
    visa_df[j]=visa_df[j].map(dict1)
visa_df

case_id  continent  education_of_employee  has_job_experience  \
0            0          0                      0                   0   
1            1          0                      1                   1   
2            2          0                      2                   0   
3            3          0                      2                   0   
4            4          1                      1                   1   
...        ...        ...                    ...                 ...   
25475    25475          0                      2                   1   
25476    25476          0                      0                   1   
25477    25477          0                      1                   1   
25478    25478          0                      1                   1   
25479    25479          0                      2                   1   

       requires_job_training  no_of_employees  yr_of_estab  \
0                          0            14513         2007   
1                          0             2412         2002   
2                          1            44444         2008   
3                          0               98         1897   
4                          0             1082         2005   
...                      ...              ...          ...   
25475                      1             2601         2008   
25476                      0             3274         2006   
25477                      0             1121         1910   
25478                      1             1918         1887   
25479                      0             3195         1960   

       region_of_employment  prevailing_wage  unit_of_wage  \
0                         0         592.2029             0   
1                         1       83425.6500             1   
2                         0      122996.8600             1   
3                         0       83434.0300             1   
4                         2      149907.3900             1   
...                     ...              ...           ...   
25475                     2       77092.5700             1   
25476                     1      279174.7900             1   
25477                     2      146298.8500             1   
25478                     0       86154.7700             1   
25479                     3       70876.9100             1   

       full_time_position  case_status  
0                       0            0  
1                       0            1  
2                       0            0  
3                       0            0  
4                       0            1  
...                   ...          ...  
25475                   0            1  
25476                   0            1  
25477                   1            1  
25478                   0            1  
25479                   0            1  

[25480 rows x 12 columns]

In [5]:
import math
x=visa_df['no_of_employees']
y=visa_df['yr_of_estab']
add1=0
add2=0
for i,j in zip(x,y):
    add1+=i
    add2+=j
numerator=add1*add2
x_var=x.var()
y_var=y.var()
denom=math.sqrt(x_var)*math.sqrt(y_var)
co_rel=numerator/denom
co_rel

7513588824.379146

In [6]:
import math
x=visa_df['no_of_employees']
y=visa_df['prevailing_wage']
add1=0
add2=0
for i,j in zip(x,y):
    add1+=i
    add2+=j
numerator=add1*add2
x_var=x.var()
y_var=y.var()
denom=math.sqrt(x_var)*math.sqrt(y_var)
co_rel=numerator/denom
co_rel

226710822.80366063

In [7]:
import math
x=visa_df['yr_of_estab']
y=visa_df['prevailing_wage']
add1=0
add2=0
for i,j in zip(x,y):
    add1+=i
    add2+=j
numerator=add1*add2
x_var=x.var()
y_var=y.var()
denom=math.sqrt(x_var)*math.sqrt(y_var)
co_rel=numerator/denom
co_rel

42760344208.735954

In [8]:
# i want to apply the map method for all categorical columns
# step-1 : we nned to get unique label
# step-2 : we need to get a number = len unique labels
# step-3 : we need to create a dictonary

In [9]:
uniq_labels=sorted(visa_df['case_status'].unique())
list1=[i for i in range(len(uniq_labels))]
dict1=dict(zip(uniq_labels,list1))
visa_df['case_status']=visa_df['case_status'].map(dict1)
visa_df

case_id  continent  education_of_employee  has_job_experience  \
0            0          0                      0                   0   
1            1          0                      1                   1   
2            2          0                      2                   0   
3            3          0                      2                   0   
4            4          1                      1                   1   
...        ...        ...                    ...                 ...   
25475    25475          0                      2                   1   
25476    25476          0                      0                   1   
25477    25477          0                      1                   1   
25478    25478          0                      1                   1   
25479    25479          0                      2                   1   

       requires_job_training  no_of_employees  yr_of_estab  \
0                          0            14513         2007   
1                          0             2412         2002   
2                          1            44444         2008   
3                          0               98         1897   
4                          0             1082         2005   
...                      ...              ...          ...   
25475                      1             2601         2008   
25476                      0             3274         2006   
25477                      0             1121         1910   
25478                      1             1918         1887   
25479                      0             3195         1960   

       region_of_employment  prevailing_wage  unit_of_wage  \
0                         0         592.2029             0   
1                         1       83425.6500             1   
2                         0      122996.8600             1   
3                         0       83434.0300             1   
4                         2      149907.3900             1   
...                     ...              ...           ...   
25475                     2       77092.5700             1   
25476                     1      279174.7900             1   
25477                     2      146298.8500             1   
25478                     0       86154.7700             1   
25479                     3       70876.9100             1   

       full_time_position  case_status  
0                       0            0  
1                       0            1  
2                       0            0  
3                       0            0  
4                       0            1  
...                   ...          ...  
25475                   0            1  
25476                   0            1  
25477                   1            1  
25478                   0            1  
25479                   0            1  

[25480 rows x 12 columns]

In [10]:
visa_df=pd.read_csv('Visadataset.csv')
for i in cat_col:
    uniq_labels=sorted(visa_df[i].unique())
    list1=[i for i in range(len(uniq_labels))]
    dict1=dict(zip(uniq_labels,list1))
    visa_df[i]=visa_df[i].map(dict1)
visa_df

case_id  continent  education_of_employee  has_job_experience  \
0            0          1                      2                   0   
1            1          1                      3                   1   
2            2          1                      0                   0   
3            3          1                      0                   0   
4            4          0                      3                   1   
...        ...        ...                    ...                 ...   
25475    17204          1                      0                   1   
25476    17205          1                      2                   1   
25477    17206          1                      3                   1   
25478    17207          1                      3                   1   
25479    17209          1                      0                   1   

       requires_job_training  no_of_employees  yr_of_estab  \
0                          0            14513         2007   
1                          0             2412         2002   
2                          1            44444         2008   
3                          0               98         1897   
4                          0             1082         2005   
...                      ...              ...          ...   
25475                      1             2601         2008   
25476                      0             3274         2006   
25477                      0             1121         1910   
25478                      1             1918         1887   
25479                      0             3195         1960   

       region_of_employment  prevailing_wage  unit_of_wage  \
0                         4         592.2029             0   
1                         2       83425.6500             3   
2                         4      122996.8600             3   
3                         4       83434.0300             3   
4                         3      149907.3900             3   
...                     ...              ...           ...   
25475                     3       77092.5700             3   
25476                     2      279174.7900             3   
25477                     3      146298.8500             3   
25478                     4       86154.7700             3   
25479                     1       70876.9100             3   

       full_time_position  case_status  
0                       1            1  
1                       1            0  
2                       1            1  
3                       1            1  
4                       1            0  
...                   ...          ...  
25475                   1            0  
25476                   1            0  
25477                   0            0  
25478                   1            0  
25479                   1            0  

[25480 rows x 12 columns]

**LabelEncoder**

- LabelEncoder is a method to convert categorical data to numerical data

- LabelEncoder is in same way how we did in above code

- the for loop map method we exploerd

- but the LabelEncoder also do the same

- It is under **sickit-learn** we usually called as **sklearn**

- under sklearn we have calss called as preprocessing

- because currently we are doing data preprocessing only

- under preprocessing calss we have a method called **LabelEncoder**

- sklearn

    - preprocessing
    
        - LabelEncoder

- any sklearn package have 3 steps

    - **step-1 :** Read the package
 
    - **step-2 :** Save the package
 
    - **step-3 :** Apply fit transform on data 

In [11]:
visa_df=pd.read_csv('Visadataset.csv')
# step-1 : read the package

from sklearn.preprocessing import LabelEncoder

# step-2 : save the package

le=LabelEncoder()

# step-3 : apply fit transform on data

visa_df['case_status']=le.fit_transform(visa_df['case_status'])
visa_df

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

       case_status  
0                1  
1                0  
2                1  
3                1  
4                0  
...            ...  
25475            0  
25476            0  
25477            0  
25478            0  
25479            0  

[25480 rows x 12 columns]

In [12]:
from sklearn.preprocessing import LabelEncoder
visa_df=pd.read_csv('Visadataset.csv')
l=LabelEncoder()
visa_df['continent']=l.fit_transform(visa_df['continent'])
visa_df

case_id  continent education_of_employee has_job_experience  \
0         EZYV01          1           High School                  N   
1         EZYV02          1              Master's                  Y   
2         EZYV03          1            Bachelor's                  N   
3         EZYV04          1            Bachelor's                  N   
4         EZYV05          0              Master's                  Y   
...          ...        ...                   ...                ...   
25475  EZYV25476          1            Bachelor's                  Y   
25476  EZYV25477          1           High School                  Y   
25477  EZYV25478          1              Master's                  Y   
25478  EZYV25479          1              Master's                  Y   
25479  EZYV25480          1            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

      case_status  
0          Denied  
1       Certified  
2          Denied  
3          Denied  
4       Certified  
...           ...  
25475   Certified  
25476   Certified  
25477   Certified  
25478   Certified  
25479   Certified  

[25480 rows x 12 columns]

In [13]:
from sklearn.preprocessing import LabelEncoder
visa_df=pd.read_csv('Visadataset.csv')
l=LabelEncoder()
for i in cat_col:
    visa_df[i]=l.fit_transform(visa_df[i])
visa_df

case_id  continent  education_of_employee  has_job_experience  \
0            0          1                      2                   0   
1            1          1                      3                   1   
2            2          1                      0                   0   
3            3          1                      0                   0   
4            4          0                      3                   1   
...        ...        ...                    ...                 ...   
25475    17204          1                      0                   1   
25476    17205          1                      2                   1   
25477    17206          1                      3                   1   
25478    17207          1                      3                   1   
25479    17209          1                      0                   1   

       requires_job_training  no_of_employees  yr_of_estab  \
0                          0            14513         2007   
1                          0             2412         2002   
2                          1            44444         2008   
3                          0               98         1897   
4                          0             1082         2005   
...                      ...              ...          ...   
25475                      1             2601         2008   
25476                      0             3274         2006   
25477                      0             1121         1910   
25478                      1             1918         1887   
25479                      0             3195         1960   

       region_of_employment  prevailing_wage  unit_of_wage  \
0                         4         592.2029             0   
1                         2       83425.6500             3   
2                         4      122996.8600             3   
3                         4       83434.0300             3   
4                         3      149907.3900             3   
...                     ...              ...           ...   
25475                     3       77092.5700             3   
25476                     2      279174.7900             3   
25477                     3      146298.8500             3   
25478                     4       86154.7700             3   
25479                     1       70876.9100             3   

       full_time_position  case_status  
0                       1            1  
1                       1            0  
2                       1            1  
3                       1            1  
4                       1            0  
...                   ...          ...  
25475                   1            0  
25476                   1            0  
25477                   0            0  
25478                   1            0  
25479                   1            0  

[25480 rows x 12 columns]

**fit and transform**

- fit means develop the application/logic

- transform means process the logic

- if we observe in map mmethod we develop dictonary logic

- we apply that dictonary is kind of fit

- developing the dictonary to column means we are transforming the data categorical to numerical

- whenever we are transforming or changing the data then use **fit transform**

- if we are not changing the data then only **fit**

**np.where**

- using np.where also we can change categorical data to numerical data

- but np.where is only for binary conditins

- np.where is same as if-else

- if condition is True then we apply true value

- if condition is false we apply false value

- so np.where use for only binary labels

- a column should have only unique label

In [14]:
visa_df=pd.read_csv('Visadataset.csv')
visa_df['case_status']=np.where(visa_df['case_status']=='Certified',1,0)
visa_df

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

       case_status  
0                0  
1                1  
2                0  
3                0  
4                1  
...            ...  
25475            1  
25476            1  
25477            1  
25478            1  
25479            1  

[25480 rows x 12 columns]

**one hot encoder**

- one hot means if one wil be on another will be off

- on represents with 1

- off represents with 0

- for example case status have 2 unique labels

    - certified
    
    - denied

- one hot encoder creates new columns which is equal to number of unique labels

- for example case status will create 2 new extra columns

    - case statua certified
    
    - case status denied

|case_status|case_status_certified|case_status_denied| 
|-|-|-|
|denied|0|1|
|certified|1|0|  
|denied|0|1|  

**pd.get_dummies**

In [16]:
visa_df=pd.read_csv('Visadataset.csv')
pd.get_dummies(data=visa_df['case_status'])

Certified  Denied
0          False    True
1           True   False
2          False    True
3          False    True
4           True   False
...          ...     ...
25475       True   False
25476       True   False
25477       True   False
25478       True   False
25479       True   False

[25480 rows x 2 columns]

In [20]:
visa_df=pd.read_csv('Visadataset.csv')
df=pd.get_dummies(data=visa_df['case_status'],dtype='int')
df['case_status']=visa_df['case_status'].values
df

Certified  Denied case_status
0              0       1      Denied
1              1       0   Certified
2              0       1      Denied
3              0       1      Denied
4              1       0   Certified
...          ...     ...         ...
25475          1       0   Certified
25476          1       0   Certified
25477          1       0   Certified
25478          1       0   Certified
25479          1       0   Certified

[25480 rows x 3 columns]

In [22]:
df1=pd.get_dummies(data=visa_df['case_status'],dtype='int')
df1

Certified  Denied
0              0       1
1              1       0
2              0       1
3              0       1
4              1       0
...          ...     ...
25475          1       0
25476          1       0
25477          1       0
25478          1       0
25479          1       0

[25480 rows x 2 columns]

In [23]:
df2=visa_df[['case_status']]
df2

case_status
0          Denied
1       Certified
2          Denied
3          Denied
4       Certified
...           ...
25475   Certified
25476   Certified
25477   Certified
25478   Certified
25479   Certified

[25480 rows x 1 columns]

**Adding the two dataframes**

- merge

- join

- concat

- append

In [25]:
pd.get_dummies(visa_df,dtype='int')

no_of_employees  yr_of_estab  prevailing_wage  case_id_EZYV01  \
0                14513         2007         592.2029               1   
1                 2412         2002       83425.6500               0   
2                44444         2008      122996.8600               0   
3                   98         1897       83434.0300               0   
4                 1082         2005      149907.3900               0   
...                ...          ...              ...             ...   
25475             2601         2008       77092.5700               0   
25476             3274         2006      279174.7900               0   
25477             1121         1910      146298.8500               0   
25478             1918         1887       86154.7700               0   
25479             3195         1960       70876.9100               0   

       case_id_EZYV02  case_id_EZYV03  case_id_EZYV04  case_id_EZYV05  \
0                   0               0               0               0   
1                   1               0               0               0   
2                   0               1               0               0   
3                   0               0               1               0   
4                   0               0               0               1   
...               ...             ...             ...             ...   
25475               0               0               0               0   
25476               0               0               0               0   
25477               0               0               0               0   
25478               0               0               0               0   
25479               0               0               0               0   

       case_id_EZYV06  case_id_EZYV07  ...  region_of_employment_South  \
0                   0               0  ...                           0   
1                   0               0  ...                           0   
2                   0               0  ...                           0   
3                   0               0  ...                           0   
4                   0               0  ...                           1   
...               ...             ...  ...                         ...   
25475               0               0  ...                           1   
25476               0               0  ...                           0   
25477               0               0  ...                           1   
25478               0               0  ...                           0   
25479               0               0  ...                           0   

       region_of_employment_West  unit_of_wage_Hour  unit_of_wage_Month  \
0                              1                  1                   0   
1                              0                  0                   0   
2                              1                  0                   0   
3                              1                  0                   0   
4                              0                  0                   0   
...                          ...                ...                 ...   
25475                          0                  0                   0   
25476                          0                  0                   0   
25477                          0                  0                   0   
25478                          1                  0                   0   
25479                          0                  0                   0   

       unit_of_wage_Week  unit_of_wage_Year  full_time_position_N  \
0                      0                  0                     0   
1                      0                  1                     0   
2                      0                  1                     0   
3                      0                  1                     0   
4                      0                  1                     0   
...                  ...                ...                  

- in one hot encoder new columns are created

- the new columns equal to number of unique labels

- because of case_status 2 extra columns are created

- because of continent 6 extra columns are created

- because of case_id have 25480 unique ids so 25480 new columns are created

In [36]:
count=0
for i in cat_col:
    count=count+visa_df[i].nunique()
print(f'without numerical column = {count}')
print(f'without numerical column = {count+len(num_col)}')

without numerical column = 25507
without numerical column = 25510


In [32]:
count=0
for i in cat_col[1:]:
    count=count+visa_df[i].nunique()
print(count)
print(count+len(num_col))

27
30


**Advantages**

- using one hot encoder we will get new columns

- the new column value at a time only one will be **true** another will be **false**

- which means new columns are **independent to each other**

- independent means no relation

- means no **multi-collineraity**

- multi-collineraity is the important property in machine learning

- **machine learning models expect the input columns should be independant to each other**

- do one hot encoder gives no multi-collineraity

**Disadvantages**

- the disadvantage is it will increase many columns

- in the above example the total columns becomes 25510

- this will happen because of case_id

- in real time we will drop the id columns

- even though we will drop we are getting 30 columns

- more columns means more time to process

- we required more infrastructure(CPU,GPU,TPU)

- this is called as **curse of dimensionality**